pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsRule.java

```diff
@@ -119,7 +119,7 @@ private boolean notInsideLoop(ASTAllocationExpression node) {
                  */
                 n = n.getParent();
             } else if (n.getParent() instanceof ASTForStatement && n.getParent().getNumChildren() > 1
-                    && n == n.getParent().getChild(1)) {
+                    && n.equals(n.getParent().getChild(1))) {
```

Comment:
I think `==` for nodes is more readable than equals. An equals calls looks like it could be recursing, because intuitively two nodes are equal if their subtree are the equal. But everywhere you replaced, we don't want to test whether the subtrees are structurally equal, we want to know whether they're the same. Only `==` captures this intent, using equals obscures this. Not to mention that equals may NPE and is in that sense less ergonomic.

I think we should enhance the rule with a list of types for which `==` is correct.